### DISCUSSION TITLE
Kriging for geotif image

### DISCUSSION THREAD

**Discussion by Empanag** [GeoStat-Framework/GSTools#199] [Q&A]
Hello, i m trying to use kriging to tiff image. the image is that below with number 1 that can be used as an array, it is regular grid. the 2nd image is the one that saga simple kriging module generates. I 'd like to produce a similar product using python. Is it possible somehow GSTools?

![image](https://user-images.githubusercontent.com/61833032/128197062-c70ba72e-b489-4404-8613-1b99bcbc3318.png)

**Comment by MuellerSeb** [comment]
Hey there,

thanks for your interest in GSTools. This is of course possible. Could you provide dummy data, so I can help you with setting up the routines?

It should look something like this in the end:
```python
import gstools as gs
import numpy as np

# your input data as x, y, data points
x, y, data = load_your_data()
cond_pos = [x, y]
mean = np.mean(data)
# simple gaussian model since your plot looks rather smooth
model = gs.Gaussian(dim=2)
krige = gs.krige.Simple(model, cond_pos, data, mean=mean, fit_variogram=True)

# output on a structured grid (image) with specified axis
xaxis, yaxis = get_your_axes()
field = krige.structured([xaxis, yaxis], return_var=False)
```

Cheers,
Sebastian

**Reply by Empanag** [reply]
@MuellerSeb  thanks for the fast response.
this the image in tiff file and i write a txt with the array. Basically the resolution of x,y is 100m.

[kriging.zip](https://github.com/GeoStat-Framework/GSTools/files/6939073/kriging.zip)

**Reply by Empanag** [reply]
I tried this code and i have this error

---------------------------------------------------------------------------
```bash
ValueError                                Traceback (most recent call last)
<ipython-input-12-369f416802a9> in <module>
     20 # simple gaussian model since your plot looks rather smooth
     21 model = gs.Gaussian(dim=2)
---> 22 krige = gs.krige.Simple(model, cond_pos, data, mean=mean, fit_variogram=True)
     23 
     24 # output on a structured grid (image) with specified axis

c:\program files\python39\lib\site-packages\gstools\krige\methods.py in __init__(self, model, cond_pos, cond_val, mean, normalizer, trend, exact, cond_err, pseudo_inv, pseudo_inv_type, fit_normalizer, fit_variogram)
    101         fit_variogram=False,
    102     ):
--> 103         super().__init__(
    104             model,
    105             cond_pos,

c:\program files\python39\lib\site-packages\gstools\krige\base.py in __init__(self, model, cond_pos, cond_val, drift_functions, ext_drift, mean, normalizer, trend, unbiased, exact, cond_err, pseudo_inv, pseudo_inv_type, fit_normalizer, fit_variogram)
    162         self._drift_functions = None
    163         self.set_drift_functions(drift_functions)
--> 164         self.set_condition(
    165             cond_pos,
    166             cond_val,

c:\program files\python39\lib\site-packages\gstools\krige\base.py in set_condition(self, cond_pos, cond_val, ext_drift, cond_err, fit_normalizer, fit_variogram)
    498             raise ValueError("Krige.set_condition: missing cond_pos/cond_val.")
    499         # correctly format cond_pos and cond_val
--> 500         self._cond_pos, self._cond_val = set_condition(
    501             cond_pos, cond_val, self.dim
    502         )

c:\program files\python39\lib\site-packages\gstools\krige\tools.py in set_condition(cond_pos, cond_val, dim)
     46     # convert the input for right shapes and dimension checks
     47     cond_val = np.array(cond_val, dtype=np.double).reshape(-1)
---> 48     cond_pos = np.array(cond_pos, dtype=np.double).reshape(dim, -1)
     49     if len(cond_pos[0]) != len(cond_val):
     50         raise ValueError(

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.
```

**Reply by MuellerSeb** [reply]
Sorry, forgot the `model`. See the updated code above.

**Reply by Empanag** [reply]
> Sorry, forgot the `model`. See the updated code above.

 this error is with model. i added model. maybe it's the array? how must be the shape of the array? mine is (68, 132)

**Reply by Empanag** [reply]
> Sorry, forgot the `model`. See the updated code above.
this is the code 
```python
filefolder="C:/Users/fermi/Desktop/SUEWS/SUEWSapp/kriging/"
rasterinput=filefolder+"T2UFfill_T2_01_14.tif"
src_ds = gdal.Open( rasterinput )
b3 = np.array(src_ds.GetRasterBand(1).ReadAsArray())


y= np.arange(322580., 335780., 100.)
x = np.arange(3906895., 3913695. , 100.)


# your input data as x, y, data points
data = b3
cond_pos = [x, y]
cond_val = [1.0]
mean = np.mean(data)
# simple gaussian model since your plot looks rather smooth
model = gs.Gaussian(dim=2)
krige = gs.krige.Simple(model, cond_pos, data, mean=mean, fit_variogram=True)

# output on a structured grid (image) with specified axis
xaxis, yaxis = get_your_axes()
field = krige.structured([xaxis, yaxis], return_var=False)
```

**Reply by MuellerSeb** [reply]
You need to use unstructured input points.
You could use `generate_grid` for that
```python
cond_pos = gs.generate_grid((x,y))
```
https://geostat-framework.readthedocs.io/projects/gstools/en/stable/tools.html#gstools.tools.generate_grid

**Reply by Empanag** [reply]
seems to work, can i take the output as an array ? the "krige" file i mean.

**Reply by MuellerSeb** [reply]
The returned `field` is a numpy array.

**Reply by Empanag** [reply]
Field is structured with the x,y coordinates. i see that need some time to get field. is there any way to get the array through krige? as it is unstructured so i can make it tiff?

**Reply by MuellerSeb** [reply]
Sorry, but I don't fully understand your question. To make it "unstructured" you could simply reshape it, couldn't you?

**Reply by Empanag** [reply]
i m running the code as it is and the `field` is running without finish the process. i ll try to explain you.

i m using two 1-D arrays for `x` and `y` and one 2-D array for `data `to run `gs.krige.Simple()`
after this i am using the  `krige.structured()` to create an image with `x` and `y` as axes 

```
# output on a structured grid (image) with specified axis
xaxis, yaxis = x, y
field = krige.structured([xaxis, yaxis], return_var=False)
```


what type of data use the `krige`? can i take the new transformed `data` from directly `krige`, a 2-D array that is the same shape with `data` but with the new values? or i have to use the `krige.structured()` to take a structure image?
i let the script run over 30 minutes for  `krige.structured()`  and haven't finished yet.

inside the file i have a tiff with the data and a txt with the array of `data` if you can used them so you can figure out or can you send me any test data to see the format?

[kriging.zip](https://github.com/GeoStat-Framework/GSTools/files/6950891/kriging.zip)

**Reply by Empanag** [reply]
@MuellerSeb i finally had some result but i have two problems,

first the execution time of `# output on a structured grid (image) with specified axis` is over 2 hours

and the result (that is below) seems very different compared to  input data and expected data that i have attached you above



![image](https://user-images.githubusercontent.com/61833032/128696795-41de64fc-48c2-45ad-84a8-a98a9bf599aa.png)

**Reply by Empanag** [reply]
i m trying to find a solution to reduce the execute time. Is there any way to reduce the range of the points that the new values is calculated from?

**Reply by MuellerSeb** [reply]
Nearest Neighboorhood Kriging or local Kriging is not available at the moment, but is on the TODO list: https://github.com/GeoStat-Framework/GSTools-Core/issues/13

One other option at the moment could be to use OpenMP (enable by default when installing via conda).
